File: src/jamyourself/countin.py
"""Percussive count-in detection.

The player counts in "1-2-3-4" percussively (claps / sticks) before every take.
Those four evenly spaced onsets give us, per take and *independently* of any
other take:
  * a hard start anchor  -> the downbeat (beat 1 of the first real bar)
  * an initial tempo      -> from the spacing of the four counts

This is the robust alternative to a cross-layer rigid offset: two different
instruments (drums vs. bass) share little spectral content, so cross-correlating
them is weak; but each take carries its own count-in, so each can be anchored to
its own musical t=0. Trim every take to its downbeat and they all start aligned.
"""
import numpy as np
import librosa

from .engine import SR, HOP, load_mono  # noqa: F401  (load_mono re-exported)

# plausible beat period for the count-in: 50..240 bpm
_MIN_IOI = 0.25
_MAX_IOI = 1.20


def _global_tempo(env, sr, hop):
    try:
        from librosa.feature.rhythm import tempo as _tempo
    except Exception:                               # older librosa
        from librosa.beat import tempo as _tempo
    return float(np.atleast_1d(
        _tempo(onset_envelope=env, sr=sr, hop_length=hop))[0])


def detect_countin(y, sr=SR, hop=HOP, search_s=8.0, n_counts=4, tol=0.08,
                   t0_window=1.6):
    """Find the percussive count-in at the start of `y`.

    The count-in shares the song tempo (you count at the song's speed), so it
    cannot be told from the music by tempo -- only by POSITION: it is the first
    regular pulse. So we:

      1. estimate the global tempo (autocorrelation over the whole take), which
         pins the beat PERIOD and resolves half/double-tempo ambiguity that a
         pure grid search falls for;
      2. for the earliest onset (the first count) find the phase: the beat period
         near the global tempo whose grid t0 + k*b lands an onset within `tol` on
         every one of the n_counts counts; among tempo octaves, prefer the one
         whose continued pulse explains the most onsets (coverage).

    Extra onsets between counts (string noise, finger taps) are tolerated.
    Returns dict: downbeat (s), bpm, beat_period (s), counts, confidence.
    Raises ValueError if no count-in is found.
    """
    env = librosa.onset.onset_strength(y=y[: int(search_s * sr)], sr=sr,
                                       hop_length=hop)
    frames = librosa.onset.onset_detect(onset_envelope=env, sr=sr,
                                        hop_length=hop, backtrack=True)
    if len(frames) < n_counts:
        raise ValueError(f"only {len(frames)} onsets found; need {n_counts}")
    times = frames * hop / sr

    b0 = 60.0 / _global_tempo(env, sr, hop)
    # tempo octaves, each searched in a narrow +/-10% band (fine phase/tempo fit)
    centers = [c for c in (b0 / 2, b0, b0 * 2) if _MIN_IOI <= c <= _MAX_IOI]

    def coverage(t0, b):
        grid = np.arange(t0, search_s, b)
        hit = sum(np.min(np.abs(times - g)) <= tol for g in grid)
        return hit

    best = None
    for t0 in times[times <= t0_window]:
        cands = []
        for c in centers:
            for b in np.arange(c * 0.9, c * 1.1, 0.003):
                grid = t0 + b * np.arange(n_counts)
                if grid[-1] > search_s:
                    break
                idx, err, ok = [], 0.0, True
                for g in grid:
                    d = np.abs(times - g)
                    j = int(d.argmin())
                    if d[j] > tol:
                        ok = False
                        break
                    idx.append(j)
                    err += d[j]
                if ok:
                    cands.append((coverage(t0, b), -err, b, idx))
        if cands:
            cands.sort(reverse=True)               # max coverage, then min err
            _, _, b, idx = cands[0]
            best = {"idx": idx, "t0": t0, "b": b}
            break                                  # earliest t0 wins
    if best is None:
        raise ValueError("no evenly-spaced count-in found")

    matched = times[best["idx"]]
    k = np.arange(n_counts)
    b_fit, a_fit = np.polyfit(k, matched, 1)        # refine period/phase
    resid = float(np.sqrt(np.mean((matched - (a_fit + b_fit * k)) ** 2)))
    return {
        "counts": matched,
        "beat_period": float(b_fit),
        "bpm": 60.0 / float(b_fit),
        "downbeat": float(a_fit + b_fit * n_counts),   # "1" after "...4"
        "confidence": float(max(0.0, 1.0 - resid / (b_fit + 1e-9))),
    }


def trim_to_downbeat(y, downbeat, sr=SR):
    """Drop everything before the downbeat so the take starts at musical t=0."""
    i = max(0, int(round(downbeat * sr)))
    return y[i:]

File: docs/js/dsp/countin.js
// Percussive count-in detection -- faithful port of the validated Python logic.
// Finds the earliest run of 4 evenly spaced onsets at a sane counting tempo.
import { SR, HOP } from "./constants.js";
import { onsetEnvelope, pickOnsets } from "./onset.js";

const MIN_IOI = 0.34; // ~176 bpm
const MAX_IOI = 0.80; // ~75 bpm  (caps below ~1s so half-tempo can't win)

function linfit(xs, ys) {
  const n = xs.length;
  let sx = 0, sy = 0, sxx = 0, sxy = 0;
  for (let i = 0; i < n; i++) { sx += xs[i]; sy += ys[i]; sxx += xs[i] * xs[i]; sxy += xs[i] * ys[i]; }
  const slope = (n * sxy - sx * sy) / (n * sxx - sx * sx);
  const intercept = (sy - slope * sx) / n;
  return { slope, intercept };
}

export function detectCountin(y, {
  sr = SR, hop = HOP, searchS = 10.0, nCounts = 4, tol = 0.08, t0Window = null,
  fromTime = 0,
} = {}) {
  if (t0Window == null) t0Window = searchS;
  const head = y.subarray ? y.subarray(0, Math.floor(searchS * sr)) : y.slice(0, Math.floor(searchS * sr));
  const env = onsetEnvelope(head, sr, hop);
  let frames = pickOnsets(env);
  // fromTime lets the user point past leading noise/fumbling ("the count-in is
  // here") — onsets before it are ignored entirely.
  if (fromTime > 0) frames = frames.filter((f) => (f * hop) / sr >= fromTime - 1e-9);
  if (frames.length < nCounts) throw new Error(`only ${frames.length} onsets; need ${nCounts}`);
  const times = frames.map((f) => (f * hop) / sr);

  const coverage = (t0, b) => {
    let hit = 0;
    for (let g = t0; g < searchS; g += b) {
      let bd = Infinity;
      for (const t of times) bd = Math.min(bd, Math.abs(t - g));
      if (bd <= tol) hit++;
    }
    return hit;
  };

  // Earliest evenly-spaced 4-grid at a sane counting tempo wins (coverage then
  // tightness break ties). Leading noise is skipped by pointing `fromTime` past
  // it (manual override in the UI), since separating arbitrary noise from a real
  // count-in by onset statistics alone isn't reliable.
  let best = null;
  for (const t0 of times) {
    if (t0 > t0Window) break;
    let t0best = null;
    for (let b = MIN_IOI; b <= MAX_IOI + 1e-9; b += 0.003) {
      const grid = [];
      for (let k = 0; k < nCounts; k++) grid.push(t0 + b * k);
      if (grid[grid.length - 1] > searchS) break;
      const idx = [];
      let err = 0, ok = true;
      for (const g of grid) {
        let bj = -1, bd = Infinity;
        for (let j = 0; j < times.length; j++) {
          const d = Math.abs(times[j] - g);
          if (d < bd) { bd = d; bj = j; }
        }
        if (bd > tol) { ok = false; break; }
        idx.push(bj); err += bd;
      }
      if (!ok) continue;
      const cov = coverage(t0, b);
      if (!t0best || cov > t0best.cov || (cov === t0best.cov && err < t0best.err))
        t0best = { idx, b, err, cov };
    }
    if (t0best) { best = t0best; break; }   // earliest valid t0
  }
  if (!best) throw new Error("no evenly-spaced count-in found");

  const matched = best.idx.map((j) => times[j]);
  const ks = matched.map((_, k) => k);
  const { slope, intercept } = linfit(ks, matched);
  let resid = 0;
  for (let k = 0; k < nCounts; k++) resid += Math.pow(matched[k] - (intercept + slope * k), 2);
  resid = Math.sqrt(resid / nCounts);
  return {
    counts: matched,
    beatPeriod: slope,
    bpm: 60 / slope,
    downbeat: intercept + slope * nCounts, // "1" after "...4"
    confidence: Math.max(0, 1 - resid / (slope + 1e-9)),
  };
}

export function trimToDownbeat(y, downbeat, sr = SR) {
  const i = Math.max(0, Math.round(downbeat * sr));
  return y.subarray ? y.subarray(i) : y.slice(i);
}
